lib/src/rules/avoid_shadowing.dart

```diff
@@ -0,0 +1,177 @@
+// Copyright (c) 2018, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library linter.src.rules.avoid_shadowing;
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:linter/src/analyzer.dart';
+
+const desc = r'Do not use a name already visible.';
+
+const details = r'''
+
+**DO** Do not use a name already visible as getter or variable.
+
+**BAD:**
+```
+var k = null;
+
+class A {
+  var a;
+}
+
+class B extends A {
+  get b => null;
+
+  m() {
+    var a; // LINT
+    var b; // LINT
+    var k; // LINT
+  }
+}
+```
+
+**GOOD:**
+```
+var k = null;
+
+class A {
+  var a;
+}
+
+class B extends A {
+  get b => null;
+
+  m() {
+    var c; // OK
+    var d; // OK
+    var e; // OK
+  }
+}
+```
+''';
+
+class AvoidShadowing extends LintRule {
+  _Visitor _visitor;
+
+  AvoidShadowing()
+      : super(
+            name: 'avoid_shadowing',
+            description: desc,
+            details: details,
+            group: Group.errors) {
+    _visitor = new _Visitor(this);
+  }
+
+  @override
+  AstVisitor getVisitor() => _visitor;
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+
+  _Visitor(this.rule);
+
+  @override
+  visitVariableDeclarationStatement(VariableDeclarationStatement node) {
+    final variables = node.variables.variables;
+
+    bool skipInstanceMembers = false;
+    AstNode current = node;
+    while (current != null) {
+      if (current is ClassDeclaration) {
+        _checkClass(current, variables, onlyStatics: skipInstanceMembers);
+      } else if (current is ConstructorDeclaration) {
+        _checkParameters(current.parameters, variables);
+        if (current.factoryKeyword != null) {
+          skipInstanceMembers = true;
+        }
+      } else if (current is MethodDeclaration) {
+        _checkParameters(current.parameters, variables);
+        if (current.isStatic) {
+          skipInstanceMembers = true;
+        }
+      } else if (current is FunctionExpression) {
+        _checkParameters(current.parameters, variables);
+      } else if (current is FunctionDeclaration) {
+        _checkParameters(current.functionExpression.parameters, variables);
+      } else if (current is FunctionDeclarationStatement) {
+        _checkParameters(
+            current.functionDeclaration.functionExpression.parameters,
+            variables);
+      }
+
+      if (current.parent is Block) {
+        _checkParentBlock(current, variables);
+      }
+
+      current = current.parent;
+      if (current == null) {
```

Comment:
The body of this `if` could be moved outside the `while`.

Reply:
done